code/modules/client/client_colour.dm

```diff
@@ -42,229 +28,247 @@
-			qdel(colour)
-			break
+	if (!istype(new_color))
+		CRASH("Invalid color type or datum for add_client_colour: [new_color ? "[new_color] ([new_color.type])" : "null"]")
 
-/**
- * Gets the resulting colour/tone from client_colours.
- * In the case of multiple colours, they'll be converted to RGBA matrices for compatibility,
- * summed together, and then each element divided by the number of matrices. (except we do this with lists because byond)
- * target is the target variable.
- */
-#define MIX_CLIENT_COLOUR(target)\
-	var/_our_colour;\
-	var/_number_colours = 0;\
-	var/_pool_closed = INFINITY;\
-	for(var/_c in client_colours){\
-		var/datum/client_colour/_colour = _c;\
-		if(_pool_closed < _colour.priority){\
-			break\
-		};\
-		_number_colours++;\
-		if(_colour.override){\
-			_pool_closed = _colour.priority\
-		};\
-		if(!_our_colour){\
-			_our_colour = _colour.colour;\
-			continue\
-		};\
-		if(_number_colours == 2){\
-			_our_colour = color_to_full_rgba_matrix(_our_colour)\
-		};\
-		var/list/_colour_matrix = color_to_full_rgba_matrix(_colour.colour);\
-		var/list/_L = _our_colour;\
-		for(var/_i in 1 to 20){\
-			_L[_i] += _colour_matrix[_i]\
-		};\
-	};\
-	if(_number_colours > 1){\
-		var/list/_L = _our_colour;\
-		for(var/_i in 1 to 20){\
-			_L[_i] /= _number_colours\
-		};\
-	};\
-	target = _our_colour\
-
-#define CLIENT_COLOR_FILTER_KEY "fake_client_color"
+	// Ensure that if a color with this source is already present, we either abort or get rid of it
+	var/datum/client_colour/existing_color = get_client_colour(source)
+	if (existing_color)
+		if (existing_color.type == new_color.type && !force)
+			return existing_color
+		qdel(existing_color)
+	client_colours[new_color] = source
+	animate_client_colour(new_color.fade_in)
+	return new_color
 
 /**
- * Resets the mob's client.color to null, and then reapplies a new color based
- * on the client_colour datums it currently has.
- */
+ * Removes a color type from a specific source from mob's client_colours list
+ * source - color source to remove
+*/
+
+/mob/proc/remove_client_colour(source)
+	var/datum/client_colour/existing_color = get_client_colour(source)
+	if (!existing_color)
+		return FALSE
+	qdel(existing_color)
+	return TRUE
+
+/mob/proc/get_client_colour(source)
+	for(var/datum/client_colour/color as anything in client_colours)
+		if (client_colours[color] == source)
+			return color
+
+/mob/proc/get_client_colour_filters()
+	. = list()
+	// sortTim sorts the passed list instead of making the copy, and so does reverse_range
+	var/list/used_colors = reverse_range(sortTim(client_colours.Copy(), GLOBAL_PROC_REF(cmp_client_colours)))
+	var/current_color = null
+	var/color_num = 0
+	var/color_prio = 1
+
+	for (var/datum/client_colour/client_color as anything in used_colors)
+		color_num += 1
+
+		var/list/filter_color = null
+		if (islist(client_color.color))
+			filter_color = client_color.color
+			if (!filter_color["type"])
+				filter_color = null
```

Comment:
What's this even for? I don't know of any `client_colour` with that kind of value for color.